crypto/rsa/rsa_ossl.c

```diff
@@ -664,10 +668,37 @@ static int rsa_ossl_mod_exp(BIGNUM *r0, const BIGNUM *I, RSA *rsa, BN_CTX *ctx)
     }
 
     if (rsa->flags & RSA_FLAG_CACHE_PUBLIC)
-        if (!BN_MONT_CTX_set_locked
-            (&rsa->_method_mod_n, rsa->lock, rsa->n, ctx))
+        if (!BN_MONT_CTX_set_locked(&rsa->_method_mod_n, rsa->lock,
+                                    rsa->n, ctx))
+            goto err;
+
+    if (smooth) {
+        if (/* m1 = I moq q */
+            !bn_from_mont_fixed_top(m1, I, rsa->_method_mod_q, ctx)
+            || !bn_to_mont_fixed_top(m1, m1, rsa->_method_mod_q, ctx)
+            /* m1 = m1^dmq1 mod q */
+            || !BN_mod_exp_mont_consttime(m1, m1, rsa->dmq1, rsa->q, ctx,
+                                          rsa->_method_mod_q)
+            /* r1 = I mod p */
+            || !bn_from_mont_fixed_top(r1, I, rsa->_method_mod_p, ctx)
+            || !bn_to_mont_fixed_top(r1, r1, rsa->_method_mod_p, ctx)
+            /* r1 = r1^dmp1 mod p */
+            || !BN_mod_exp_mont_consttime(r1, r1, rsa->dmp1, rsa->p, ctx,
+                                          rsa->_method_mod_p)
```

Comment:
Just in case question is popped. Outputs from BN_mod_exp_mont_consttime are not fixed-length, but they go straight to length-invariant bn_mod_sub_fixed_top, so that only variation would be bn_correct_top at the ends of BN_mod_exp_mont_consttime, Variation is very unlike and it's not unreasonable to assume that even them it won't be measurable in the context.

Reply:
It occurs to me that, unlike the issue below with `r0`, this `bn_correct_top` is not obviously saved by blinding because the distribution involves `p` and `q` separately.

I would have considered the `bn_correct_top` one to clear out, but I also favor a more complete non-minimal `BIGNUM` story, which would allow `BN_mod_exp_mont_consttime` to just return non-minimal values. How much work to put into the interim fix, I don't have strong views on.

Reply:
I don't quite follow. (It's late here:-) You mean to introduce bn_mod_exp_fixed_length? It's not very much trouble...